zookeeper/source/zk_nn_client.cc

```diff
@@ -689,54 +817,94 @@ namespace zkclient{
-                }
-                if (datanodes.size() == replication_factor) {
-                    LOG(INFO) << CLASS_NAME << "Found " << replication_factor << " datanodes";
-                    break;
+                    bool exclude = false;
+                    if (excluded_datanodes.size() > 0) {
+                        // Remove the excluded datanodes from the live list
+                        std::vector<std::string>::iterator it = std::find(
+                                excluded_datanodes.begin(),
+                                excluded_datanodes.end(),
+                                datanode);
+
+                        if (it == excluded_datanodes.end()) {
+                            // This datanode was not in the excluded list, so keep it
+                            exclude = false;
+                        } else {
+                            exclude = true;
+                        }
+                    }
+
+                    if (!exclude) {
+                        std::string dn_stats_path = HEALTH_BACKSLASH + datanode + STATS;
+                        std::vector<uint8_t> stats_payload;
+                        stats_payload.resize(sizeof(DataNodePayload));
+                        if (!zk->get(dn_stats_path, stats_payload, error_code, sizeof(DataNodePayload))) {
+                            LOG(ERROR) << CLASS_NAME << "Failed to get " << dn_stats_path;
+                            return false;
+                        }
+                        DataNodePayload stats = DataNodePayload();
+                        memcpy(&stats, &stats_payload[0], sizeof(DataNodePayload));
+                        LOG(INFO) << "\t DN stats - free_bytes: " << unsigned(stats.free_bytes);
+                        if (stats.free_bytes > blocksize) {
+							auto queue = REPLICATE_QUEUES + datanode;
+							auto repl_item = util::concat_path(queue, std::to_string(blockId));
+							bool alreadyOnQueue;
+
+							if (zk->exists(repl_item, alreadyOnQueue, error_code)) {
+								if (alreadyOnQueue) {
+                                    LOG(INFO) << "Skipping target" << datanode;
+                                    continue;
+								}
+							}
+                            LOG(INFO) << "Pushed target: " << datanode << " with " << stats.xmits << " xmits";
+                            targets.push(TargetDN(datanode, stats.free_bytes, stats.xmits));
+                        }
+                    }
                 }
             }
+
+            LOG(INFO) << "There are " << targets.size() << " viable DNs. Requested: " << replication_factor;
+            if (targets.size() < replication_factor) {
+                LOG(ERROR) << "Not enough available DNs! Available: " << targets.size() << " Requested: " << replication_factor;
+                // no return because we still want the client to write to some datanodes
+            }
+
+            while (datanodes.size() < replication_factor && targets.size() > 0) {
```

Comment:
hm this code should be on a branch somewhere. let me look for it, if not i will rewrite.